File: frontend/src/api/printRoom.ts
import http from './http'

export async function issueOfficialContract(payload: {
  work_order_id: number
  contract_no: string
  remark?: string
}) {
  const { data } = await http.post('/print-room/issue-official-contract', payload)
  return data as { message: string }
}

export interface PrintRoomInfoItem {
  work_order_id: number
  current_status?: string | null
  current_status_display?: string | null
  contract_no?: string | null
  paper_report_no?: string | null
  copy_count?: number | null
  formal_report_count?: number | null
  remark?: string | null
}

export async function getPrintRoomInfo(workOrderId: number) {
  const { data } = await http.get(`/print-room/work-orders/${workOrderId}`)
  return data as PrintRoomInfoItem
}

export interface PrintRoomContractParticipant {
  user_id?: number | null
  user_name?: string | null
}

export interface PrintRoomContractFileItem {
  id: number
  origin_file_name: string
  file_size?: number | null
  uploaded_at?: string | null
  uploaded_by_user_id?: number | null
  uploaded_by_user_name?: string | null
  is_current: boolean
}

export interface ContractPrintRoomInfoItem {
  work_order_id: number
  current_status?: string | null
  current_status_display?: string | null
  project_leader: PrintRoomContractParticipant
  contract_reviewer: PrintRoomContractParticipant
  print_room_handler: PrintRoomContractParticipant
  original_contract_files: PrintRoomContractFileItem[]
  stamped_contract_scan_files: PrintRoomContractFileItem[]
  can_upload_scan: boolean
  can_send_to_project_leader: boolean
  can_return_to_print_room: boolean
  can_confirm_complete: boolean
}

export async function getContractPrintRoomInfo(workOrderId: number) {
  const { data } = await http.get(`/print-room/contracts/work-orders/${workOrderId}`)
  return data as ContractPrintRoomInfoItem
}

export async function sendContractToProjectLeader(payload: { work_order_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/contracts/send-to-project-leader', payload)
  return data as { message: string }
}

export async function returnContractToPrintRoom(payload: { work_order_id: number; remark: string }) {
  const { data } = await http.post('/print-room/contracts/return-to-print-room', payload)
  return data as { message: string }
}

export async function confirmContractComplete(payload: { work_order_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/contracts/confirm-complete', payload)
  return data as { message: string }
}

export async function transferPrintRoom(payload: { work_order_id: number; handler_user_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/transfer-print-room', payload)
  return data as { message: string }
}

export async function rollbackThird(payload: { work_order_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/rollback-third', payload)
  return data as { message: string }
}

export async function markContractError(payload: { work_order_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/contract-error', payload)
  return data as { message: string }
}

export async function reportError(payload: { work_order_id: number; remark?: string }) {
  const { data } = await http.post('/print-room/report-error', payload)
  return data as { message: string }
}

export interface PrintRoomRecordItem {
  id: number
  work_order_id: number
  handled_by: number
  paper_report_no: string
  copy_count: number
  printed_at?: string
  remark?: string
}

export async function issuePaperReport(payload: {
  work_order_id: number
  paper_report_no: string
  copy_count: number
  remark?: string
}) {
  const { data } = await http.post('/print-room/issue-paper-report', payload)
  return data as PrintRoomRecordItem
}

File: backend/app/schemas/print_room.py
from datetime import datetime

from pydantic import BaseModel


class IssueOfficialContractRequest(BaseModel):
    work_order_id: int
    contract_no: str
    remark: str | None = None


class IssuePaperReportRequest(BaseModel):
    work_order_id: int
    paper_report_no: str
    copy_count: int = 1
    remark: str | None = None


class PrintRoomInfoResponse(BaseModel):
    work_order_id: int
    current_status: str | None = None
    current_status_display: str | None = None
    print_room_status: str | None = None
    print_room_status_display: str | None = None
    contract_no: str | None = None
    paper_report_no: str | None = None
    copy_count: int | None = None
    formal_report_count: int | None = None
    remark: str | None = None


class PrintRoomContractParticipant(BaseModel):
    user_id: int | None = None
    user_name: str | None = None


class PrintRoomContractFileResponse(BaseModel):
    id: int
    origin_file_name: str
    file_size: int | None = None
    uploaded_at: datetime | None = None
    uploaded_by_user_id: int | None = None
    uploaded_by_user_name: str | None = None
    is_current: bool = True


class ContractPrintRoomInfoResponse(BaseModel):
    work_order_id: int
    current_status: str | None = None
    current_status_display: str | None = None
    project_leader: PrintRoomContractParticipant
    contract_reviewer: PrintRoomContractParticipant
    print_room_handler: PrintRoomContractParticipant
    original_contract_files: list[PrintRoomContractFileResponse] = []
    stamped_contract_scan_files: list[PrintRoomContractFileResponse] = []
    can_upload_scan: bool = False
    can_send_to_project_leader: bool = False
    can_return_to_print_room: bool = False
    can_confirm_complete: bool = False


class PrintRoomRollbackRequest(BaseModel):
    work_order_id: int
    remark: str | None = None


class TransferPrintRoomRequest(BaseModel):
    work_order_id: int
    handler_user_id: int
    remark: str | None = None


class PrintRoomRecordResponse(BaseModel):
    id: int
    work_order_id: int
    handled_by: int
    paper_report_no: str
    copy_count: int
    printed_at: datetime | None
    remark: str | None
